plugins/formal-verification/src/org/jetbrains/kotlin/formver/conversion/StmtConversionVisitor.kt

```diff
@@ -114,23 +114,38 @@ class StmtConversionVisitor : FirVisitor<Exp, StmtConversionContext>() {
 
     override fun visitFunctionCall(functionCall: FirFunctionCall, data: StmtConversionContext): Exp {
         val id = functionCall.calleeReference.toResolvedCallableSymbol()!!.callableId
-        // TODO: figure out a more structured way of doing this
-        if (id.packageName.asString() == "kotlin.contracts" && id.callableName.asString() == "contract") return UnitDomain.element
-        return when (functionCall.dispatchReceiver) {
-            is FirNoReceiverExpression -> {
-                val symbol = functionCall.calleeReference.resolved!!.resolvedSymbol as FirNamedFunctionSymbol
-                val calleeSig = data.add(symbol)
-                val args = functionCall.argumentList.arguments.map {
-                    it.accept(this, data)
-                }
-                val returnVar = data.newAnonVar(calleeSig.returnType)
-                val returnExp = returnVar.toLocalVar()
-                data.addDeclaration(returnVar.toLocalVarDecl())
-                data.addStatement(Stmt.MethodCall(calleeSig.name.asString, args, listOf(returnExp)))
-                returnExp
-            }
-            else -> TODO("Implement function call visitation with receiver")
+        val specialFunc = SpecialFunctions.byCallableId[id]
+        if (specialFunc != null) {
+            if (specialFunc !is SpecialFunctionImplementation) return UnitDomain.element
+        }
+
+        val args = getFunctionCallArguments(functionCall).map { it.accept(this, data) }
+
+        if (specialFunc != null) {
+            if (specialFunc !is SpecialFunctionImplementation) return UnitDomain.element
+            return specialFunc.convertCall(args, data)
+        }
```

Comment:
what about replacing nested if statements with this: `(specialFunc as? SpecialFunctionImplementation)?.let { return specialFunc.convertCall(args, data) }`
because I think that `return UnitDomain.element` can never be reached

Reply:
It can't, but unfortunately Kotlin can't deduce that :(
That said, I think we may be able to use contracts to make this nicer, give me a moment :)

Reply:
I tried it and it compiles

Reply:
because I think that the case `specialFunc != null && specialFunc !is SpecialFunctionImplementation` is captured before and even if the compiler is not able to deduce it, I want to return `return specialFunc.convertCall(args, data)` only in the case `specialFunc` is an instance of `SpecialFunctionImplementation`

Reply:
I mean, ideally we'd want the compiler to remember that `specialFunc != null` implies `specialFunc !is SpecialFunctionImplementation`.  But it doesn't do that, so we need to do something else.  Your suggestion compiles, but I think there may be a clearer way to put it... I'll think about it.